File: src/main/java/club/xiaojiawei/strategy/AbstractPhaseStrategy.java
package club.xiaojiawei.strategy;

import club.xiaojiawei.bean.entity.Card;
import club.xiaojiawei.bean.entity.ExtraEntity;
import club.xiaojiawei.bean.entity.TagChangeEntity;
import club.xiaojiawei.enums.ConfigurationEnum;
import club.xiaojiawei.enums.DeckEnum;
import club.xiaojiawei.enums.StepEnum;
import club.xiaojiawei.listener.log.PowerLogListener;
import club.xiaojiawei.status.War;
import club.xiaojiawei.utils.GameUtil;
import club.xiaojiawei.utils.PowerLogUtil;
import club.xiaojiawei.utils.SystemUtil;
import javafx.beans.property.BooleanProperty;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

import javax.annotation.Resource;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.List;
import java.util.Objects;
import java.util.Properties;
import java.util.concurrent.atomic.AtomicReference;

import static club.xiaojiawei.data.ScriptStaticData.*;

/**
 * @author 肖嘉威
 * @date 2022/11/26 17:59
 */
@Slf4j
public abstract class AbstractPhaseStrategy{
    @Resource
    protected PowerLogListener powerLogListener;
    @Resource
    protected GameUtil gameUtil;
    @Resource
    protected AtomicReference<BooleanProperty> isPause;
    @Resource
    protected Properties scriptConfiguration;
    /**
     * 告诉Power.log监听器，AbstractPhaseStrategy是否正在处理日志
     */
    @Getter
    private volatile static boolean dealing;
    private String lastDiscoverEntityId;

    public void deal(String line) {
        dealing = true;
        try{
            beforeDeal();
            dealLog(line);
            afterDeal();
        }finally {
            dealing = false;
        }
    }
    private void dealLog(String line){
        RandomAccessFile accessFile = powerLogListener.getAccessFile();
        long mark;
        while (!isPause.get().get()) {
            try {
                if (line == null) {
                    mark = accessFile.getFilePointer();
                    SystemUtil.delay(1000);
                    if (accessFile.length() <= mark && War.getMe() != null){
                        List<Card> cards = War.getMe().getSetasideArea().getCards();
                        int size = cards.size();
                        if (
                                size >= 3
                                && !Objects.equals(lastDiscoverEntityId, cards.get(cards.size() - 1).getEntityId())
                                && Objects.equals(cards.get(size - 1).getCreator(), cards.get(size - 2).getCreator())
                                && Objects.equals(cards.get(size - 1).getCreator(), cards.get(size - 3).getCreator())
                        ){
                            log.info("触发发现动作");
                            lastDiscoverEntityId = cards.get(cards.size() - 1).getEntityId();
                            DeckEnum.valueOf(scriptConfiguration.getProperty(ConfigurationEnum.DECK.getKey())).getAbstractDeckStrategy().discoverChooseCard(cards.get(size - 3), cards.get(size - 2), cards.get(size - 1));
                        }
                    }
                }else if (powerLogListener.isRelevance(line)){
                    if (log.isDebugEnabled()){
                        log.debug(line);
                    }
                    if (line.contains(TAG_CHANGE)){
                        if (dealTagChangeThenIsOver(line, PowerLogUtil.dealTagChange(line)) || War.getCurrentTurnStep() == StepEnum.FINAL_GAMEOVER){
                            break;
                        }
                    }else if (line.contains(SHOW_ENTITY)){
                        if (dealShowEntityThenIsOver(line, PowerLogUtil.dealShowEntity(line, accessFile))){
                            break;
                        }
                    }else if (line.contains(FULL_ENTITY)){
                        if (dealFullEntityThenIsOver(line, PowerLogUtil.dealFullEntity(line, accessFile))){
                            break;
                        }
                    }else if (line.contains(CHANGE_ENTITY)){
                        if (dealChangeEntityThenIsOver(line, PowerLogUtil.dealChangeEntity(line, accessFile))){
                            break;
                        }
                    }else {
                        if (dealOtherThenIsOver(line)){
                            break;
                        }
                    }
                }
                line = accessFile.readLine();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }
    protected void beforeDeal(){
        log.info("当前处于：" + War.getCurrentPhase().getComment());
    }
    protected void afterDeal(){
        log.info(War.getCurrentPhase().getComment() + " -> 结束");
    }

    protected boolean dealTagChangeThenIsOver(String line, TagChangeEntity tagChangeEntity){
        return false;
    }
    protected boolean dealShowEntityThenIsOver(String line, ExtraEntity extraEntity){
        return false;
    }
    protected boolean dealFullEntityThenIsOver(String line, ExtraEntity extraEntity){
        return false;
    }
    protected boolean dealChangeEntityThenIsOver(String line, ExtraEntity extraEntity){
        return false;
    }
    protected boolean dealOtherThenIsOver(String line){
        return false;
    }
}

File: Hearthstone-Script/src/main/java/club/xiaojiawei/hsscript/strategy/AbstractPhaseStrategy.kt
package club.xiaojiawei.hsscript.strategy

import club.xiaojiawei.bean.Card
import club.xiaojiawei.bean.isValid
import club.xiaojiawei.config.log
import club.xiaojiawei.enums.StepEnum
import club.xiaojiawei.enums.WarPhaseEnum
import club.xiaojiawei.hsscript.bean.DeckStrategyThread
import club.xiaojiawei.hsscript.bean.log.ExtraEntity
import club.xiaojiawei.hsscript.bean.log.TagChangeEntity
import club.xiaojiawei.hsscript.consts.CHANGE_ENTITY
import club.xiaojiawei.hsscript.consts.FULL_ENTITY
import club.xiaojiawei.hsscript.consts.SHOW_ENTITY
import club.xiaojiawei.hsscript.consts.TAG_CHANGE
import club.xiaojiawei.hsscript.listener.log.PowerLogListener
import club.xiaojiawei.hsscript.status.PauseStatus
import club.xiaojiawei.hsscript.strategy.DeckStrategyActuator.discoverChooseCard
import club.xiaojiawei.hsscript.utils.PowerLogUtil.dealChangeEntity
import club.xiaojiawei.hsscript.utils.PowerLogUtil.dealFullEntity
import club.xiaojiawei.hsscript.utils.PowerLogUtil.dealShowEntity
import club.xiaojiawei.hsscript.utils.PowerLogUtil.dealTagChange
import club.xiaojiawei.hsscript.utils.PowerLogUtil.isRelevance
import club.xiaojiawei.hsscript.utils.SystemUtil
import club.xiaojiawei.interfaces.PhaseStrategy
import club.xiaojiawei.status.War.currentPhase
import club.xiaojiawei.status.War.currentTurnStep
import club.xiaojiawei.status.War.me
import club.xiaojiawei.util.isTrue
import java.io.IOException

/**
 * 游戏阶段抽象类
 * @author 肖嘉威
 * @date 2022/11/26 17:59
 */
abstract class AbstractPhaseStrategy : PhaseStrategy {

    private var lastDiscoverEntityId: String? = null

    override fun deal(line: String) {
        dealing = true
        try {
            beforeDeal()
            dealLog(line)
            afterDeal()
        } finally {
            dealing = false
        }
    }

    private fun dealLog(line: String) {
        val accessFile = PowerLogListener.logFile
        accessFile ?: return
        var l:String? = line
        var mark: Long
        while (!PauseStatus.isPause) {
            try {
                if (l == null) {
                    mark = accessFile.filePointer
                    SystemUtil.delay(1000)
                    if (accessFile.length() <= mark && me.isValid()) {
                        val cards: List<Card> = me.setasideArea.cards
                        val size = cards.size
                        if (size >= 3 && lastDiscoverEntityId != cards.last().entityId && cards[size - 1].creator == cards[size - 2].creator
                            && cards[size - 1].creator == cards[size - 3].creator
                        ) {
                            lastDiscoverEntityId = cards.last().entityId
                            if (currentPhase != WarPhaseEnum.REPLACE_CARD) {
                                log.info { "触发发现动作" }
                                (DeckStrategyThread({
                                    discoverChooseCard(
                                        cards[size - 3],
                                        cards[size - 2],
                                        cards[size - 1]
                                    )
                                }, "Discover Choose Card Thread").also { addTask(it) }).start()

                            }
                        }
                    }
                } else if (isRelevance(l)) {
                    log.debug { l }
                    if (l.contains(TAG_CHANGE)) {
                        if (dealTagChangeThenIsOver(
                                l,
                                dealTagChange(l)
                            ) || currentTurnStep == StepEnum.FINAL_GAMEOVER
                        ) {
                            break
                        }
                    } else if (l.contains(SHOW_ENTITY)) {
                        if (dealShowEntityThenIsOver(l, dealShowEntity(l, accessFile))) {
                            break
                        }
                    } else if (l.contains(FULL_ENTITY)) {
                        if (dealFullEntityThenIsOver(l, dealFullEntity(l, accessFile))) {
                            break
                        }
                    } else if (l.contains(CHANGE_ENTITY)) {
                        if (dealChangeEntityThenIsOver(l, dealChangeEntity(l, accessFile))) {
                            break
                        }
                    } else {
                        if (dealOtherThenIsOver(l)) {
                            break
                        }
                    }
                }
                l = accessFile.readLine()
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }

    protected fun beforeDeal() {
        WarPhaseEnum.find(this)?.let {
            log.info { "当前处于：" + it.comment }
        }
    }

    protected fun afterDeal() {
        WarPhaseEnum.find(this)?.let {
            log.info { it.comment + " -> 结束" }
        }
    }

    protected open fun dealTagChangeThenIsOver(line: String, tagChangeEntity: TagChangeEntity): Boolean {
        return false
    }

    protected open fun dealShowEntityThenIsOver(line: String, extraEntity: ExtraEntity): Boolean {
        return false
    }

    protected open fun dealFullEntityThenIsOver(line: String, extraEntity: ExtraEntity): Boolean {
        return false
    }

    protected open fun dealChangeEntityThenIsOver(line: String, extraEntity: ExtraEntity): Boolean {
        return false
    }

    protected open fun dealOtherThenIsOver(line: String): Boolean {
        return false
    }

    companion object{
        var dealing = false
        private val tasks:MutableList<Thread> = mutableListOf()

        fun addTask(task:Thread) {
            tasks.add(task)
        }

        fun cancelAllTask(){
            val toList = tasks.toList()
            tasks.clear()
            toList.forEach {
                it.isAlive.isTrue {
                    it.interrupt()
                }
            }
        }
    }

}
